diff --git a/plugins/TagManager/config/config.php b/plugins/TagManager/config/config.php
index 676c3d0..69289f4 100644
--- a/plugins/TagManager/config/config.php
+++ b/plugins/TagManager/config/config.php
@@ -3,12 +3,12 @@
 return array(
     'TagManagerContainerStorageDir' => function () {
         // the location where we store the generated javascript or json container files
-        return '/js';
+        return '/../../../../var/lib/matomo/tagmanager';
     },
     'TagManagerContainerWebDir' => function (\Psr\Container\ContainerInterface $c) {
         // the path under which the containers are available through the web. this may be different to the storage
         // path if using eg htaccess rewrites
-        return $c->get('TagManagerContainerStorageDir');
+        return '/js/tagmanager';
     },
     'TagManagerContainerFilesPrefix' => function () {
         // the prefix for any container file
