--- a/tools/frr-reload.py	1970-01-01 01:00:01.000000000 +0100
+++ b/tools/frr-reload.py	2024-06-14 13:44:48.061087035 +0200
@@ -1869,7 +1869,10 @@

     elif args.reload:
         if not os.path.isdir("/var/log/frr/"):
-            os.makedirs("/var/log/frr/", mode=0o0755)
+            try:
+                os.makedirs("/var/log/frr/", mode=0o0755)
+            except FileExistsError:
+                pass

         logging.basicConfig(
             filename="/var/log/frr/frr-reload.log",
