diff --git a/src/global/signal_handler.h b/src/global/signal_handler.h
index d768e4e47c..1e33077b2a 100644
--- a/src/global/signal_handler.h
+++ b/src/global/signal_handler.h
@@ -22,11 +22,7 @@

 typedef void (*signal_handler_t)(int);

-#ifndef HAVE_REENTRANT_STRSIGNAL
-# define sig_str(signum) sys_siglist[signum]
-#else
-# define sig_str(signum) strsignal(signum)
-#endif
+#define sig_str(signum) strsignal(signum)

 void install_sighandler(int signum, signal_handler_t handler, int flags);
